share/doc/gtools.1.md
% GTOOLS(1) gtools 0.3.0 | User Commands
% Frank Scheiner
% Mar 23, 2017


# NAME #

**gtools**


# SYNOPSIS #

**gtools [function [arguments]...]** or

**function [arguments]...** if symlinked


# DESCRIPTION #

**gtools** is a multi-call shell script that combines various GridFTP
functionality (provided by **uberftp(1c)**) into a single executable. Most
people will create a link to gtools for each function they wish to use and
gtools will act like whatever is was invoked as.


# COMMON OPTIONS #

All **gtools** functions provide a terse runtime description of their behavior
when invoked without arguments.

# FUNCTIONS #

Currently defined functions:

(g)cat, (g)ls, (g)mkdir, (g)mv, (g)rm

All functions support host aliases and remote directory browsing via bash
completion. Function names are without the **g[...]** prefix when used as
arguments of **gtools**. When used directly (via symlinks) the **g[...]** prefix
should be used to differentiate them from the similar OS tools (**cat(1)**,
**ls(1)**, **mkdir(1)**, **mv(1)** and **rm(1)**).

# FUNCTION DESCRIPTIONS #

## (g)cat ##

**gcat _url_**

Print to stdout the contents of the remote file given in _url_ (like **cat [...]**).

## (g)ls ##

**gls _url_**

List the remote file or the contents of the remote directory given in _url_.
Output is similar to **ls -la [...]**.

## (g)mkdir ##

**gmkdir _url_**

Create the remote directory given in _url_. Creates non-existing directories
recursively (like **mkdir -p [...]**) but needs one GridFTP operation per directory!

## (g)mv ##

**gmv _url1_ _url2_**

Move/rename the remote file or directory in _url1_ to the given path and/or
name in _url2_. Only works when both _url1_ and _url2_ point to the same remote
GridFTP service!

## (g)rm ##

**grm _url_**

Remove the remote file (like **rm [...]**) or empty directory (like **rmdir [...]**) given in _url_.


# SEE ALSO #

**uberftp(1c)**, **halias(1)**, **cat(1)**, **ls(1)**, **mkdir(1)**, **rm(1)**, **rmdir(1)**


ChangeLog
2017-03-27  Frank Scheiner <scheiner@hlrs.de>

* 0.8.0 (gtransfer toolkit):

- New functionality

The (guc) sync level used by gtransfer for transfers can now be selected by
users. Syncing can also be disabled.

The host aliases can now be retrieved and installed from a file provided
locally or from a remote repository, just like dpaths and dparams.

Wrappers for uberftp (called gtools) were added that provide shorthand
functionality to users for common operations on remote GridFTP servers like
(g)cat, (g)ls, (g)mkdir, (g)mv and (g)rm. All wrappers support host aliases and
remote directory browsing just like gt.

2016-09-08  Frank Scheiner <scheiner@hlrs.de>

*0.7.1 (gtransfer toolkit):

- Fixes

Due to an erroneous comparison, data transfers with URLs that contained
host aliases that weren't prefixed with usernames failed. This because
a wrong replacement in the GridFTP URLs was applied due to the
erroneous comparison. This is now fixed both in gtransfer and the
gtransfer bash completion.

2016-09-01  Frank Scheiner <scheiner@hlrs.de>

*0.7.0 (gtransfer toolkit):

- New functionality

Gt now also supports "<USER>@" as prefix to host aliases. This was
already included in the Bash completion but not activated until now.

- Fixes

"helperFunctions/getURLWithoutPath()" did not anticipate user names in
front of host addresses in GridFTP URLs. This is now corrected.

Until now gt overwrote log files from unsuccessful tries and retries of
a transfer. This was a source of irritation during evaluation of
transfers as the start and stop timestamps did not resemble the start
and stop of the whole transfer. And it also meant a loss of possibly
helpful info. This was now changed so that older logfiles of the same
transfer are preserved and newer ones get an additional identifier for
each retry attached to their names.

2016-06-30  Frank Scheiner <scheiner@hlrs.de>

*0.6.0 (gtransfer toolkit):

- New functionality

The stall timeout setting for globus-url-copy is now configurable via
the gtransfer configuration file or from a dparam. If not configured in
a dparam, the setting from the gtransfer configuration file is used. The
old hard-coded default of 30 seconds is not necessarily suitable for
data transfers from sites which first need to stage in files from e.g.
tape before a data transfer can begin (e.g. sites using HPSS).

2016-04-12  Frank Scheiner <scheiner@hlrs.de>

*0.5.1 (gtransfer toolkit):

- Fixes

Using a "/" or "/*" in a source URL to transfer all files in a given dir
now no longer makes problems for "stat" if only one file existed in the
given dir. This fixes issue #41.

Special chars in URLs need to be percent or URL encoded when provided to
globus-url-copy on the commandline. This means that currently URLs with
special chars also have to be provided to gt percent or URL encoded.
This is now also mentioned in the gt manpage.
Before using stat, the path and file name in the source URL are now
decoded beforehand, which makes the call to stat also working for
URLs with special chars. This fixes issue #42.

2016-04-12  Frank Scheiner <scheiner@hlrs.de>

*0.5.0 (gtransfer toolkit):

- New functionality

Gt options can now be used multiple times on the command line, the last
setting of an option wins. This allows for setting gt default options
via shell aliases.

- Fixes

Added missing gt options to bash completion.

Enhanced the workaround for the guc bug (no size string in created
transfer lists if only a single file is to be transferred) to get the
size for `file://` source URLs with the `stat` tool locally.

2016-01-27  Frank Scheiner <scheiner@hlrs.de>

*0.4.1 (gtransfer toolkit):

- Fixes

Multipathing: A wrong variable name was corrected which prevented correct
functioning of multipathing when using a transfer list as input to gt.

Host aliases: Until now when trying to dealias host aliases that cannot be
accessed by the effective user, the halias tool didn't exit a while loop and
hence looped until interrupted by the user. This is now corrected.

2015-09-17  Frank Scheiner <scheiner@hlrs.de>

*0.4.0 (gtransfer toolkit):

- New functionality

Users can now enable data channel checksumming or data channel encryption with a
gtransfer option.

- Fixes

PID functionality: The rule for resolving a PID to an URL was adapted to backend
changes in EUDAT.

2015-04-23  Frank Scheiner <scheiner@hlrs.de>

*0.3.0 (gtransfer dist):

- New functionality

Data transfer using multipathing: Gtransfer can now distribute a data transfer
over multiple paths. This way users can benefit from the combined bandwidth of
multiple paths. For more details please have a look into the [README].

[README]: https://github.com/fr4nk5ch31n3r/gtransfer#data-transfer-using-multipathing

- Fixes

Included a workaround for an old `globus-url-copy` (guc) [bug]: A transfer list
created for a single file does not contain the size of the file. gt now
determines the file size in this case with `uberftp` and updates the transfer
list. The file size is needed for calculating the transfer rate and for the
bin packing heuristics used by the multipathing functionality.

[bug]: https://globus.atlassian.net/browse/GT-318

For native OS packages (like RPMs) all tools now use `/usr/share/gtransfer` as
base path for bash libraries. This fixes [issue #26].

Integrated a 1 second timeout to the ping command in
`helperFunctions/isValidUrl()` to keep wait time low when transferring between
hosts not answering echo requests.

[issue #26]: https://github.com/fr4nk5ch31n3r/gtransfer/issues/26

2013-10-14  Frank Scheiner  <frank.scheiner@web.de>

gtransfer dist v0.2.1:

*0.1.1 (halias):
Fixed [issue #21]. This fix corrects a wrong variable name in halias that
prevented correct functioning after doing a system installation.

[issue #21]: https://github.com/fr4nk5ch31n3r/gtransfer/issues/21


2013-08-23  Frank Scheiner  <frank.scheiner@web.de>

gtransfer dist v0.2.0:

*0.2.0 (gtransfer):
Added the following functionality:


etc/bash_completion.d/gtools.sh
:<<COPYRIGHT

Copyright (C) 2016-2017 Frank Scheiner, HLRS, Universitaet Stuttgart

The program is distributed under the terms of the GNU General Public License

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

This product includes software developed by members of the DEISA project
www.deisa.org. DEISA is an EU FP7 integrated infrastructure initiative under
contract number RI-222919.

COPYRIGHT

gtools_complete_urls()
{
	########################################################################
	#  HELPER FUNCTIONS ####################################################
	########################################################################
	getPathFromURL()
	{
		#  determines the path portion from the URL:
		#
		#  gsiftp://venus.milkyway.universe:2811(/path/to/)file
		#
		#  usage:
		#+ getPathFromURL "URL"

		local URL="$1"

		#local path=$(echo "$URL" | sed -e "s|$( getURLWithoutPath $URL )||")
		#  local path?
		if echo $URL | grep "^.*://" &>/dev/null; then
			#  no
			local tmp=$( echo "$URL" | cut -d '/' -f '4-' )
			#  add leading '/'
			tmp="/$tmp"
		else
			#  yes
			tmp=$URL
		fi

		#  strip any file portion from path
		path=$(echo $tmp | grep -o '/.*/')
		if [[ $path == "" ]]; then
			path="/"
		fi	

		echo "$path"
	}

	getPathFromAliasUrl()
	{
		local url="$1"
		
		# if there's currentl only the alias, assume "/"
		if ! echo "$url" | grep "/" &>/dev/null; then
			echo "/"
			return
		fi
		
		local path=${url#*/}
		
		#  strip any file portion from path
		#path=${path%/*}
		# add leading '/'
                path="/$path"
		#  strip any file portion from path
		path=$(echo $path | grep -o '/.*/')
		
		#if [[ "$path" != "" ]]; then
		#	echo "/$path/"
		#else
		#	echo "/"
		#fi
		if [[ $path == "" ]]; then
			path="/"
		fi	
		
		echo "$path"
	}

	getURLWithoutPath()
	{
		#  determines the URL portion that consists of the protocol id, the
		#+ domain name and the port, or "file://":
		#
		#  (gsiftp://venus.milkyway.universe:2811)/path/to/file
		#  (file://)/path/to/local/file
		#
		#  usage:
		#+ getURLWithoutPath "URL"

		local URL="$1"
	
		#  TODO:
		#+ support URLs not containing any port descriptions:
		#
		#  done!
	
		:<<-COMMENT
		from: <http://wiki.linuxquestions.org/wiki/Regular_expression>
		"
		echo gsiftp://venus.milkyway.universe/path/to/file | sed "s;\(gsiftp://[^/]*\)/.*;\1;"
		"

		or

		"
		echo gsiftp://venus.milkyway.universe/path/to/file | cut -d '/' -f "1-3"
		"

		returns:
		"
		gsiftp://venus.milkyway.universe
		"
		COMMENT

		#local tmp=$(echo "$URL" | grep -o "gsiftp://.*:[[:digit:]]*")
		local tmp=""
		#  URL starting with "/", then this is a local path (equal to
		#+  "file://$URL".
		if [[ ${URL:0:1} == "/" ]]; then
			#echo "DEBUG: 1"
			tmp="file://"
		#  valid URL
		else
			#echo "DEBUG: 3"
			tmp=$( echo $URL | cut -d '/' -f "1-3" )
		fi
		#if [[ "$tmp" == "" ]]; then
		#	tmp=$( echo "$URL" | grep -o "file://" )
		#fi

		#  does $tmp start with 'gsiftp://'?
		if echo $tmp | grep '^gsiftp://' &>/dev/null; then
			#  if yes, check if port is provided
			if echo $tmp | grep -o ':[[:digit:]].*' &>/dev/null; then
				#  port provided by user, don't modify string
				:
			else
				#  no port provided, add default gsiftp port
				tmp="${tmp}:2811"
			fi
		#  does $tmp start with 'ftp://'?
		elif echo $tmp | grep '^ftp://' &>/dev/null; then
			#  if yes, check if port is provided
			if echo $tmp | grep -o ':[[:digit:]].*' &>/dev/null; then
				#  port provided by user, don't modify string
				:
			else
				#  no port provided, add default ftp port
				tmp="${tmp}:21"
			fi
		#  does $tmp start with 'http://'?
		elif echo $tmp | grep '^http://' &>/dev/null; then
			#  if yes, check if port is provided
			if echo $tmp | grep -o ':[[:digit:]].*' &>/dev/null; then
				#  port provided by user, don't modify string
				:
			else
				#  no port provided, add default ftp port
				tmp="${tmp}:80"
			fi
		#  does $tmp start with 'https://'?
		elif echo $tmp | grep '^https://' &>/dev/null; then
			#  if yes, check if port is provided
			if echo $tmp | grep -o ':[[:digit:]].*' &>/dev/null; then
				#  port provided by user, don't modify string
				:
			else
				#  no port provided, add default ftp port
				tmp="${tmp}:443"
			fi
		fi

		local URLWithoutPath=$tmp

		echo "$URLWithoutPath"
	}
	########################################################################

	local cur prev opts

	#  defuse ":" in completions, as the ":" implies specific readline
	#+ behaviour.
	COMP_WORDBREAKS=${COMP_WORDBREAKS//:}
	#  also defuse "@" (used when usernames are provided in URLs)
	COMP_WORDBREAKS=${COMP_WORDBREAKS//@}

	COMPREPLY=()

	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"

	if hash globus-url-copy &>/dev/null; then
		#  complete remote paths
		if echo "$cur" | grep '^gsiftp://.*:.*' &>/dev/null; then

			userhost=$( getURLWithoutPath "${cur}" )
			#echo -e "\n$userhost" 1>&2
			userpath=$( getPathFromURL "${cur}" )
			#echo "$userpath" 1>&2

			local remote_paths=( $( globus-url-copy -list "${userhost}${userpath}*" | sed -e 's/^\ *//' -e 1d ) )

			local remote_urls=$( for path in "${remote_paths[@]}"; do echo ${userhost}${userpath}${path}; done )
		
			COMPREPLY=( $(compgen -W "${remote_urls}" -- ${cur}) )
			return 0
		
		elif echo "$cur" | grep '^ftp://.*:.*' &>/dev/null; then
	
			userhost=$( getURLWithoutPath "${cur}" )
			userpath=$( getPathFromURL "${cur}" )
		
			local remote_paths=( $( globus-url-copy -list "${userhost}${userpath}*" | sed -e 's/^\ *//' -e 1d ) )
		
			local remote_urls=$( for path in "${remote_paths[@]}"; do echo ${userhost}${userpath}${path}; done )
		
			COMPREPLY=( $(compgen -W "${remote_urls}" -- ${cur}) )
			return 0
		fi

		if hash halias &>/dev/null; then
		
			alias="${cur%%/*}" ## remove path
			user="${alias%%@*}"
			alias="${alias#*@}" ## remove "user@"
		
			if halias --is-alias "$alias" &>/dev/null; then						
	
				userhost=$( halias --dealias "$alias" )

				if [[ "$user" != "$alias" ]]; then

					userhost=${userhost/:\/\//:\/\/$user@}
				fi
				#echo -e "\n$userhost A$alias U$user" 1>&2
				userpath=$( getPathFromAliasUrl "$cur" )
				#echo "$userpath" 1>&2
		
				local remote_paths=( $( globus-url-copy -list "${userhost}${userpath}*" | sed -e 's/^\ *//' -e 1d ) )
		
				if [[ "$user" != "$alias" ]]; then
					local remote_urls=$( for path in "${remote_paths[@]}"; do echo ${user}@${alias}${userpath}${path}; done )
				else
					local remote_urls=$( for path in "${remote_paths[@]}"; do echo ${alias}${userpath}${path}; done )
				fi
			
				COMPREPLY=( $(compgen -W "${remote_urls}" -- ${cur}) )
				return 0
			fi
		fi
	fi

	#  only complete source URL host parts if dpath is available
	if hash dpath &>/dev/null; then
		#  complete source URL host parts
		local sites=$( dpath --list-sources )
	else
		local sites=""
	fi
	
	if hash halias &>/dev/null; then
		local aliases=$( halias --list )
	else
		local aliases=""
	fi
	
	if [[ "$sites" != "" && "$aliases" != "" ]]; then
		COMPREPLY=( $(compgen -W "${sites} ${aliases}" -- ${cur}) )
	elif [[ "$sites" != "" ]]; then
		COMPREPLY=( $(compgen -W "${sites}" -- ${cur}) )
	elif [[ "$aliases" != "" ]]; then
		COMPREPLY=( $(compgen -W "${aliases}" -- ${cur}) )
	fi

	return 0
}


gtools_complete_general_options()
{
	local cur prev opts

	#  defuse ":" in completions, as the ":" implies specific readline
	#+ behaviour.
	COMP_WORDBREAKS=${COMP_WORDBREAKS//:}
	#  also defuse "@" (used when usernames are provided in URLs)
	COMP_WORDBREAKS=${COMP_WORDBREAKS//@}

	COMPREPLY=()

	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"

	COMPREPLY=( $( compgen -W '--help -h --version -V' -- "$cur" ) )
        COMPREPLY=( "${COMPREPLY[@]/%/ }" )
        return 0
}


gtools_complete_functions()
{
	local cur prev opts

	#  defuse ":" in completions, as the ":" implies specific readline
	#+ behaviour.
	COMP_WORDBREAKS=${COMP_WORDBREAKS//:}
	#  also defuse "@" (used when usernames are provided in URLs)
	COMP_WORDBREAKS=${COMP_WORDBREAKS//@}

	COMPREPLY=()

	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"

	COMPREPLY=( $( compgen -W "gcat gls gmkdir gmv grm" -- ${cur} ) )
	return 0
}


_gtools()
{
	local cur prev opts

	#  defuse ":" in completions, as the ":" implies specific readline
	#+ behaviour.
	COMP_WORDBREAKS=${COMP_WORDBREAKS//:}
	#  also defuse "@" (used when usernames are provided in URLs)
	COMP_WORDBREAKS=${COMP_WORDBREAKS//@}

	COMPREPLY=()

	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"

	#echo "${#COMP_WORDS[@]}" 1>&2

	if [[ $( basename $prev) == "gtools" ]]; then

		gtools_complete_functions
		return 0
	fi

	if [[ $( basename $prev ) == "gcat" || \
	      $( basename $prev ) == "gls" || \
	      $( basename $prev ) == "gmkdir" || \
	      $( basename $prev ) == "grm" ]]; then

		case $cur in
		    -*)
		        gtools_complete_general_options
		        return 0
		        ;;
		    *)
		        #+ pass through
		        ;;
		esac

		gtools_complete_urls
		return 0

	elif [[ $( basename ${COMP_WORDS[0]} ) == "gmv" ]]; then

		case $cur in
		    -*)
		        gtools_complete_general_options
		        return 0
		        ;;
		    *)
		        #+ pass through
		        ;;
		esac

		if [[ ${#COMP_WORDS[@]} -le 3 ]]; then

			gtools_complete_urls
			return 0
		fi

	elif [[ $( basename ${COMP_WORDS[1]} ) == "gmv" ]]; then

		case $cur in
		    -*)
		        gtools_complete_general_options
		        return 0
		        ;;
		    *)
		        #+ pass through
		        ;;
		esac

		if [[ ${#COMP_WORDS[@]} -le 4 ]]; then

			gtools_complete_urls
			return 0
		fi
	fi	

	return 0
}

complete -o nospace -F _gtools gtools gcat gls gmkdir gmv grm

